File: track_b_systems/python_core/week2/day10/system.py
import os
import json

MAX_LIMIT=25000
DAILY_LIMIT=50000
DATA_FILE="data.json"

def load_user():
	if not os.path.exists(DATA_FILE):
		print("System data file missing.")
		return None
	
	else :
		with open (DATA_FILE,"r") as f:
			return json.load(f)
		
def save_user(user_data):
	with open (DATA_FILE,"w") as f:
		json.dump(user_data,f,indent=4)

def check_balance(user):
	print("Balance : ",user['balance'])

def deposit_amount(user):
	dep_amount=int(input("Enter the amount to be depsited : \n"))
	if dep_amount<=0:
		print("Invalid Amount. Deosit failed !\n")

	else :
		user['balance']+=dep_amount
		print("Amount added successfully : ",dep_amount)

def withdraw_amount(user):
	wit_amount=int(input("Enter the amount to be withdrawn : \n"))

	if wit_amount>user['balance'] :
		print("Insuffecient Amount. Withdraw Failed !\n")

	elif wit_amount>MAX_LIMIT :
		print("Max transaction limit reached.\n")

	elif wit_amount+user['daily_withdrawn']>DAILY_LIMIT :
		print("Daily limit reached. Try again tommorow !")
	
	else :
		user['balance']-=wit_amount
		user['daily_withdrawn']+=wit_amount
		print("Amount Withdrawn Sucessfully :",wit_amount)

def authentication(user):
	name=str(input("Enter the username : \n"))
	if name==user['username']:
		for i in range(3,0,-1):
			password=str(input("Enter the password : \n"))
			if password==user['pin'] :
				print(f"System Logged in Sucessfully.\n")
				return True
			
			else :
				print(f"Wrong Pin. {i-1} chances reamining ! !\n")

	else: 
			print(f"Wrong Username.\n")

	return False
		

def main():
	user=load_user()
	if not user:
		return
	
	system_on=False
	system_on=authentication(user)
	while system_on==True:
		print("\n------- SYSTEM CONTROL PANEL -------\n")
		print(f"\n------- User : {user['username']} -------\n")
		print("1. Check Balance Amount\n")
		print("2. Deposit Amount\n")
		print("3. Withdraw Amount\n")
		print("4. Exit System\n")

		choice=int(input("Enter the choice : \n"))
		if choice==1:
			check_balance(user)
		
		elif choice==2:
			deposit_amount(user)
		
		elif choice==3:
			withdraw_amount(user)

		elif choice==4:
			print("System shutting down...")
			break

		else:
			print("invalid Choice !")


if __name__=="__main__":
	main()

File: track_a_engineering/cpp_dsa/week2/day10/system.cpp
#include <iostream>
#include <string>
using namespace std;

//----- Config Layer -----//

const int max_limit=25000;
const int daily_limit=50000;
	
//----- Data Model -----//

struct User
{
	string username;
	string pin;
	int balance;
	int daily_withdrawn;
};

//----- Auth Layer -----//

bool authentication(User &u)
{
	string password;
	string username;
	cout<<"Enter the username : \n";
	cin>>username;

	for(int i=3;i>=1;i--)
	{
		cout<<"Enter the password : \n";
		cin>>password;

		if(password==u.pin)
		{
			cout<<"Correct Password.\n"<< username <<" Logged in!\n";
			return true;
		}

		else
		{
			cout<<"Wrong Password. Chances Remaining : "<<i-1<<endl;
		}
	}
	return false;
	
}

//------ UI Layer  -----//

void show_menu()
{
	cout<<"\n------ SYSTEM CONTROL PANEL -----\n";
	cout<<"1. Check Balance \n";
	cout<<"2. Deposit Amount \n";
	cout<<"3. Withdraw Amount \n";
	cout<<"4. Exit System \n";
}

//----- Business Logic-----//

void check_balance(User &u)
{
	cout<<"Balance : \n"<<u.balance<<endl;
}

void deposit_amount(User &u)
{
	int amount;
	cout<<"Enter the amount to be added : \n";
	cin>>amount;

	if(amount>0)
	{
		u.balance+=amount;
		cout<<"Deposited : \n"<<amount<<endl;
	}

	else
	{
		cout<<"Invalid Amount. \n";
	}
}

void withdraw_amount(User &u)
{
	int withdraw;
	cout<<"Enter the amount to be withdrawn : \n";
	cin>>withdraw;

	if(withdraw<=0)
	{
		cout<<"Invalid amount.\n";
	}
	
	else if(withdraw>u.balance)
	{
		cout<<"Insuffecient Balance.\n";
	}

	else if(withdraw>max_limit)
	{
		cout<<"Max Transaction Limit Exceeded.\n";
	}

	else if(u.daily_withdrawn+withdraw>daily_limit)
	{
		cout<<"Daily Limit Reached.Try again Tommorow!\n";
	}

	else
	{
		u.balance-=withdraw;
		u.daily_withdrawn+=withdraw;
		cout<<"Amount withdrawn sucessfully : \n"<<withdraw<<endl;
	}

}

//----- Controller -----//

void controller(User &u)
{
	bool system_on=true;

	while(system_on)
	{
		show_menu();

		int choice;
		cout<<"Enter the choice : \n";
		cin>>choice;
		switch (choice)
		{
		case 1:	
			check_balance(u);
			break;
			
		case 2:
			deposit_amount(u);
			break;

		case 3:
			withdraw_amount(u);
			break;

		case 4:
			cout<<"System shutting down...\n";
			system_on=false;
			break;

		default:
			cout<<"Invalid Option.\n";
		}
	}
}

//----- MAIN ------//

int main()
{
	User user;
	user.username="Raghu";
	user.balance=1000;
	user.daily_withdrawn=0;
	user.pin="sarghi919";

	if(authentication(user)==true)
	{
		controller(user);
	}

	else
	{
		cout<<"System Locked.\n";
	}
	return 0;
}